src/shared/cpu-set-util.c

```diff
@@ -199,3 +200,31 @@ int parse_cpu_set_extend(
 
         return cpu_set_add_all(old, old_allocated, cpuset, allocated);
 }
+
+void cpu_set_to_dbus(cpu_set_t *set, size_t allocated) {
+        size_t len = allocated / sizeof(unsigned long);
+
+        assert(set);
+        assert(allocated >= sizeof(unsigned long));
+
+        if (sizeof(uint64_t) == sizeof(unsigned long))
+                swap_array_htole64((uint64_t*) set->__bits, len);
+        else if (sizeof(uint32_t) == sizeof(unsigned long))
+                swap_array_htole32((uint32_t*) set->__bits, len);
+        else
+                assert_not_reached("Unknown size of unsigned long");
+}
+
+void cpu_set_from_dbus(cpu_set_t *set, size_t allocated) {
+        size_t len = allocated / sizeof(unsigned long);
+
+        assert(set);
+        assert(allocated >= sizeof(unsigned long));
+
+        if (sizeof(uint64_t) == sizeof(unsigned long))
+                swap_array_le64toh((uint64_t*) set->__bits, len);
+        else if (sizeof(uint32_t) == sizeof(unsigned long))
```

Comment:
Hmm, I've checked glibc and musl and they both use unsigned long int. Also libc can't just change the type since it would ABI break as well as API break in a sense that internal type is exposed via documentation [1].

[1] http://man7.org/linux/man-pages/man3/CPU_SET.3.html

```
Since CPU sets are bit masks allocated in units of long words, the
actual number of CPUs in a dynamically allocated CPU set will be
rounded up to the next multiple of sizeof(unsigned long).  An
application should consider the contents of these extra bits to be
undefined.
```

IOW, I don't think the current code is going to be a problem any time soon. However, I can rewrite it if you feel strongly about it.

Reply:
Ok, so I did some more digging and found this glibc bug,

https://sourceware.org/bugzilla/show_bug.cgi?id=19313

According to the bug the internal type should be long long on 32bit Intel, hence the bit width should be 64bit everywhere (most likely because this is what kernel expects userspace to pass-in for sched_setaffinity()). IOW, maybe we could rely on this and simplify the code by assuming that internal storage must have 64bit width.

Reply:
Please don't make those assumptions. There is really no need to.